### ISSUE TITLE
desire to create associative array with anonymous domain

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#11346]
I'd like to create an associative array with an "anonymous" domain,
the same way one might declare `var B:[1..0] int`.

However the following leads to an error that mentions chpl_ensureDomainExpr.

``` chapel
var A:[domain(string)] int;

A["hi"] = 1;

writeln(A);
```

Should this pattern be allowed?

**Comment by mppf** [comment]
If the pattern is not allowed, it seems strange, because `domain(string)` ought to be usable in a "type context" and I'd expect the `:[ ... ]` in the array declaration to be a "type context".

**Comment by bradcray** [comment]
If we support something like this, I don't think this is the right syntax.  Though we are in a type context in that we're between the `:` and `=` in the declaration context, the domain expression itself is (traditionally) a value expression.  So I would expect something more like `var A: [new domain(string)] int;` to be the way to write it.  Or to invent some sugar for "empty domain of strings" though I've never had good ideas around that...

That said, I'm also wondering more and more whether `var B: [1..0] int;` should really be supported as it is today, particularly if/when we split array-as-vec functionality into its own type (issue #9452).  Specifically it seems to cause more problems and confusion than value, and also seems a bit odd in the sense of "This appears to be a param range/domain value—what permits me to modify it later?"  (in asking that, I'm anticipating the const-domain optimizations that we need in other contexts like #9414.  B's domain appears to be a literal const/param domain, so what permits us to change it?

So I wonder whether maybe in both cases what we really want is "vector" and "set" types that are built in terms of arrays and domains within, but don't expose those capabilities to the (typical) user who doesn't want to have to think about domains.

(For reference, the case that bit me this week when using `var B: [1..0] int;` was declaring a type of the form:

```chapel
record R {
  var B: [1..0] int;
}
```

where things like the default copy initializer "don't work" because it ends up assigning arrays of different sizes.  Whereas if I wrote the declaration like this:

```chapel
record R {
  const BDom = {1..0};
  var B: [BDom] int;
}
```

or this:

```chapel
record R {
  var myVec: vector;  // where vector is essentially like the record just above
}
```

then it worked because the assignment of the domain field first caused the array field to be the proper size before it was assigned.

**Comment by mppf** [comment]
> So I wonder whether maybe in both cases what we really want is "vector" and "set" types that are built in terms of arrays and domains within, but don't expose those capabilities to the (typical) user who doesn't want to have to think about domains.

I think that's a reasonable reaction here. I'd generally be happy to have "vector" "set" and "map" types that behave more like these types in other languages. Domains and arrays are cool, but they are also a bit foreign to people coming from Python or C++.

**Comment by mppf** [comment]
One more thought, I wonder if having `map` and `set` would allow us to use standard library implementations of the reduction one might want to use when creating per-thread maps or per-thread sets.

**Comment by bradcray** [comment]
Right, I think it makes sense to increasingly think of domains and arrays as "the core things upon which the language is based" but that if we create the right abstractions, more and more users may not ever need to touch them for common cases ("I want to create a matrix, vector, set, block-distributed array").  I think the belief is that we need interfaces to make this fly though, from a code reuse perspective (i.e., where today we might write things to take "an array" we'd write them to take something supporting "iteration" or "random access" or whatever).

Going back, though, I'm realizing that if we supported `new domain(string)` in the `[...]` context, that doesn't raise the same concerns I have with `[1..0]` since the former says "give me a new variable of this type" without running into the "this seems to be a fixed domain that can't be modified."  So that reaction only applies to my thought to have a sugar for "give me an empty associative domain of strings".
